src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -413,90 +417,130 @@ internal static void WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<Re
+        // From https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-readfilescatter:
+        // "The file handle must be created with [...] the FILE_FLAG_OVERLAPPED and FILE_FLAG_NO_BUFFERING flags."
+        private static bool CanUseScatterGatherWindowsAPIs(SafeFileHandle handle)
+            => handle.IsAsync && ((handle.GetFileOptions() & SafeFileHandle.NoBuffering) != 0);
+
+        // From the same source:
+        // "Each buffer must be at least the size of a system memory page and must be aligned on a system
+        // memory page size boundary. The system reads/writes one system memory page of data into/from each buffer."
+        // This method returns true if the buffers can be used by
+        // the Windows scatter/gather API, which happens when they are:
+        // 1. aligned at page size boundaries
+        // 2. exactly one page long each (our own requirement to prevent partial reads)
+        // 3. not bigger than 2^32 - 1 in total
+        // This function is also responsible for pinning the buffers if they
+        // are suitable and they must be unpinned after the I/O operation completes.
+        // It also returns a pointer with the segments to be passed to the
+        // Windows API, and the total size of the buffers that is needed as well.
+        // The pinned MemoryHandles and the pointer to the segments must be cleaned-up
+        // with the CleanupScatterGatherBuffers method.
+        private static unsafe bool TryPrepareScatterGatherBuffers<T, THandler>(IReadOnlyList<T> buffers,
+            THandler handler, out MemoryHandle[] handlesToDispose, out IntPtr segmentsPtr, out int totalBytes)
+            where THandler: struct, IMemoryHandler<T>
         {
-            if (!handle.IsAsync)
-            {
-                return ScheduleSyncReadScatterAtOffsetAsync(handle, buffers, fileOffset, cancellationToken);
-            }
+            int pageSize = s_cachedPageSize;
+            Debug.Assert(BitOperations.IsPow2(pageSize), "Page size is not a power of two.");
+            // We take advantage of the fact that the page size is
+            // a power of two to avoid an expensive modulo operation.
+            long alignedAtPageSizeMask = pageSize - 1;
+
+            int buffersCount = buffers.Count;
+            handlesToDispose = new MemoryHandle[buffersCount];
+            segmentsPtr = IntPtr.Zero;
+            totalBytes = 0;
+
+            // "The array must contain enough elements to store nNumberOfBytesToWrite bytes of data, and one element for the terminating NULL. "
+            long* segments = (long*) NativeMemory.Alloc((nuint)(buffersCount + 1), sizeof(long));
+            segments[buffersCount] = 0;
 
-            if (CanUseScatterGatherWindowsAPIs(handle))
+            bool success = false;
+            try
             {
-                long totalBytes = 0;
-                int buffersCount = buffers.Count;
+                long totalBytes64 = 0;
```

Comment:
Nit: why the `64` suffix?  We don't typically do that.

Reply:
There is an `out int totalBytes` argument. We want to make sure it does not overflow.

Reply:
Ok